File: packages/audio/android/src/main/java/dev/tcode/thinmpr/audio/service/SongService.kt
package dev.tcode.thinmpr.audio.service

import android.content.Context
import dev.tcode.thinmpr.audio.extension.toMap
import dev.tcode.thinmpr.audio.model.valueObject.AlbumId
import dev.tcode.thinmpr.audio.model.valueObject.ArtistId
import dev.tcode.thinmpr.audio.model.valueObject.SongId
import dev.tcode.thinmpr.audio.repository.SongRepository
import dev.tcode.thinmpr.audio.service.contract.SongServiceContract

class SongService(context: Context): SongServiceContract {
    private val repository = SongRepository(context)

    override fun getAllSongs(): List<Map<String, Any>> {
        val songs = repository.findAll()

        return songs.map { it.toMap() }
    }

    override fun getSongsByArtistId(artistId: String): List<Map<String, Any>> {
        val songs = repository.findByArtistId(ArtistId(artistId))

        return songs.map { it.toMap() }
    }

    override fun getSongsByIds(ids: List<String>): List<Map<String, Any>> {
        val songIds = ids.map { SongId(it) }
        val songs = repository.findByIds(songIds)

        return songs.map { it.toMap() }
    }

    override fun getSongsByAlbumId(albumId: String): List<Map<String, Any>> {
        val songs = repository.findByAlbumId(AlbumId(albumId))

        return songs.map { it.toMap() }
    }

    override fun getSongById(id: String): Map<String, Any>? {
        val songId = SongId(id)
        val song = repository.findById(songId) ?: return null

        return song.toMap()
    }
}

File: packages/audio/ios/Service/SongService.swift
class SongService: SongServiceContract {
    private let albumRepository = AlbumRepository()
    private let songRepository = SongRepository()

    func getAllSongs() -> [[String: Any]] {
        let songs = songRepository.findAll()

        return songs.map { $0.toMap() }
    }

    func getSongsByAlbumId(albumId: String) -> [[String: Any]] {
        let songs = songRepository.findByAlbumId(albumId: AlbumId(id: albumId))

        return songs.map { $0.toMap() }
    }

    func getSongsByArtistId(artistId: String) -> [[String: Any]] {
        let albums = albumRepository.findByArtistId(
            artistId: ArtistId(id: artistId)
        )
        let albumIds = albums.map { $0.id }
        let songs = songRepository.findByAlbumIds(albumIds: albumIds)

        return songs.map { $0.toMap() }
    }

    func getSongById(id: String) -> [String: Any]? {
        guard let song = songRepository.findBySongId(songId: SongId(id: id))
        else {
            return nil
        }

        return song.toMap()
    }

    func getSongsByIds(ids: [String]) -> [[String: Any]] {
        let songIds = ids.map { SongId(id: $0) }
        let songs = songRepository.findBySongIds(songIds: songIds)

        return songs.map { $0.toMap() }
    }
}
